app/experimenter/docs/adrs/0002-use-react-bootstrap-cra-gql-for-nimbus.md

```diff
@@ -0,0 +1,113 @@
+
+With the decision made to build the Nimbus console as part of the Experimenter codebase ([see ADR #0001](https://github.com/mozilla/experimenter/blob/main/app/experimenter/docs/adrs/0001-build-nimbus-console-in-experimenter.md)), it was determined we likely want a "new front-end." This ADR, while slightly atypical as it does not serve to document a single architectural change, outlines a proposed directory tree change regarding the front-end, a proposed new front-end stack, how it compares to the current `rapid` stack, and how we will benefit from these changes.
+
+## Decision Drivers
+
+Drivers for directory tree changes:
+* A more obvious separation of concerns for:
+  * the front-end and back-end
+  * legacy front-end code and Nimbus front-end code
+
+Drivers for the proposed new stack:
+* Better development tooling
+* A more cohesive fit into the FxA team's stack and processes
+* Development speed
+
+Drivers for starting a new application:
+* Time needed and ease of refactoring the current `rapid` React application vs time needed to start with a new application
+
+## Proposed Changes
+
+### App Directory Tree
+
+Currently, a `static/` directory contains the front-end applications with some shared assets, and a `templates/` directory lives adjacent to it.  To provide a more obvious separation of concerns in the directory tree:
+
+* Rename `static/` to `legacy-ui/` since `static/rapid/` will become obsolete and `static/core/` will be the legacy Normandy platform
+* Move `templates/` into `legacy-ui/`
+* Use `create-react-app` to create a new React application in `app/experimenter/nimbus-ui/`
+
+
+### Front-End Stack
+
+* Continue using React, but:
+  * Use the `create-react-app` toolchain
+  * Use Typescript over Proptypes for more strict typing
+  * Use GraphQL for API calls and `apollo-client-cache` for global state management
+  * Use `reach-router` instead of `react-router`
+  * Use directories for components instead of filenames
+    * The Storybook and test file will live adjacent to the component `index.tsx` file. This allows all associated component files to live in the same directory and makes imports shorter.
+* Continue using Bootstrap (`react-bootstrap`), but:
+  * Add stylelint for SCSS linting overrides
+  * Use breakpoints established in FxA
+* Utilize Storybook for browsing the UI in various states without running the stack
+* Add Sentry to catch client-side errors
+
+## Considered Alternatives
+
+* Refactor existing `rapid` project rather than use CRA for `nimbus-ui`
+* Use existing API calls instead of switching to GQL
+* Use Tailwind instead of Bootstrap
+
+### Positive Consequences
+
+Continuing to use React and Bootstrap while starting a new app with `create-react-app` and GraphQL/Apollo allows us to benefit from stack and tooling improvements while avoiding an entire rewrite of all front-end components.
```

Comment:
Worth noting that [some](https://github.com/mozilla/experimenter/blob/main/app/experimenter/static/core/js/components/MessageForm.js#L29:L29) of the React is written with `react-bootstrap` while [some](https://github.com/mozilla/experimenter/blob/main/app/experimenter/static/rapid/components/experiments/ExperimentDetails.tsx#L198:L198) is written with Bootstrap CSS. This may impact component re-usability depending on desired code consistency.

Reply:
Huh, so actually the `core` project uses `react-bootstrap` while the `rapid` project uses Bootstrap CSS. I should better differentiate the two so good call out.

Reply:
We had an issue for starting to use `react-boostrap` (https://github.com/mozilla/experimenter/issues/2942) but it ended up getting blocked on a dependency issue on `react-bootstrap`'s side (@tiftran or @jaredlockhart do you happen to have an issue reference on react-bootstrap for this?)

Reply:
Yeah teh problem was with their dependency on popper and some TS bug it had:

https://github.com/mozilla/experimenter/pull/3208#issuecomment-673050527

I was waiting for them to fix it on their side, not sure if it's been resolved yet I don't have a link handy to the issue on their side.

Reply:
Might be this:

https://github.com/popperjs/react-popper/issues/373

Reply:
If we didn't use `react-bootstrap` wouldn't we miss out on a lot of benefits like what's in https://react-bootstrap.github.io/components/forms/, or maybe we could just use those as references to build our own? I don't think using the last-working stable version then bumping is too horrible considering this issue seems fairly new, unless this is something you've ran into multiple times.

Reply:
No I mean you can just use bootstrap in the old fashioned way and say stuff like

```html
<div className="form">
  <div className="form-group">
```

rather than the react-bootstrap way of

```jsx
<Form>
  <FormGroup>
```

you lose the nice auto complete and properties and stuff, but it's basically 1:1 to convert them from one to the other when they fix the ts bug in popper.

Reply:
Or maybe it's already fixed 🤞

Reply:
> I was waiting for them to fix it on their side, not sure if it's been resolved yet I don't have a link handy to the issue on their side.

That PR confuses me. I commented in it to see if it's fixed or not. Though it looks like at time of open the latest version was 1.3.7, but they're on 2.2.3 now.

Reply:
Oh wait... they [dropped popper](https://github.com/react-bootstrap/react-bootstrap/commit/5bcfff2e1a6c8658f49be9c439e8dc447bef96b5).